File: WalletTracker/app/src/main/java/com/chtima/wallettracker/vm/CategoryViewModel.java
package com.chtima.wallettracker.vm;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;

import com.chtima.wallettracker.dao.repositories.CategoryRepository;
import com.chtima.wallettracker.models.Category;
import com.chtima.wallettracker.models.CategoryWithTransactions;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;

public class CategoryViewModel extends AndroidViewModel {

    private final CategoryRepository repository;
    private final LiveData<List<Category>> categoriesLiveData;
    private final LiveData<List<CategoryWithTransactions>> categoriesWithTransactionsLiveData;
    private LiveData<List<CategoryWithTransactions>> categoriesWithTransactionsByUserIdLiveData;

    private long userId;

    public CategoryViewModel(@NonNull Application application) {
        super(application);
        repository = new CategoryRepository(application);
        categoriesLiveData = LiveDataReactiveStreams.fromPublisher(repository.getAll());
        categoriesWithTransactionsLiveData = LiveDataReactiveStreams.fromPublisher(repository.getCategoriesWithTransactions());
        categoriesWithTransactionsByUserIdLiveData = null;
    }

    public Completable insertAll(Category...categories){
        return repository.insertAll(categories);
    }

    public LiveData<List<Category>> getAll() {
        return categoriesLiveData;
    }

    public LiveData<List<Category>> getByType(Category.CategoryType categoryType) {
        if(categoryType == null) return this.categoriesLiveData;
        return LiveDataReactiveStreams.fromPublisher(repository.getByType(categoryType));
    }

    public LiveData<List<CategoryWithTransactions>> getCategoriesWithTransactions() {
        return categoriesWithTransactionsLiveData;
    }

    public LiveData<List<CategoryWithTransactions>> getCategoriesWithTransactionsByUserId(long userId) {

        if(categoriesWithTransactionsByUserIdLiveData == null || this.userId != userId){
            this.userId = userId;
            categoriesWithTransactionsByUserIdLiveData = LiveDataReactiveStreams.fromPublisher(repository.getCategoriesWithTransactionsByUserId(userId));
        }

        return categoriesWithTransactionsByUserIdLiveData;
    }
}

File: WalletTracker/app/src/main/java/com/chtima/wallettracker/viewModels/CategoryViewModel.kt
package com.chtima.wallettracker.viewModels

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.toLiveData
import com.chtima.wallettracker.db.DatabaseUpdatedEvent
import com.chtima.wallettracker.db.repositories.CategoryRepository
import com.chtima.wallettracker.models.AppConstants
import com.chtima.wallettracker.models.Category
import com.chtima.wallettracker.models.CategoryWithTransactions
import com.chtima.wallettracker.models.SharedPreferencesKeys
import io.reactivex.rxjava3.core.Completable
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe


class CategoryViewModel(private val app: Application) : AndroidViewModel(app) {

    private val categoryRepository: CategoryRepository = CategoryRepository(app)
    private var categoriesWithTransactionsByUserLiveData : LiveData<List<CategoryWithTransactions>>? = null

    private var userID: Long = -1;

    fun insertAll(vararg categories: Category): Completable {
        return categoryRepository.insertAll(*categories)
    }

    fun getByType(type: Category.CategoryType?):LiveData<List<Category>>{
        if(type == null) return categoryRepository.getAll().toLiveData()
        return categoryRepository.getByType(type).toLiveData()
    }

    fun getCategoriesWithTransactionsByUser(): LiveData<List<CategoryWithTransactions>>{
        val userIdNow = SharedPreferencesKeys.getSharedPreferences(app).getLong(AppConstants.SELECTED_USER_ID, -1)
        if(categoriesWithTransactionsByUserLiveData == null || userID != userIdNow){
            userID = userIdNow
            categoriesWithTransactionsByUserLiveData = categoryRepository.getCategoriesWithTransactionsByUserId(userID).toLiveData()
            Log.d("CategoryViewModel", "create:LiveData")
        }

        Log.d("CategoryViewModel", "userID:$userID")

        return this.categoriesWithTransactionsByUserLiveData!!
    }

    override fun onCleared() {
        super.onCleared()
        EventBus.getDefault().unregister(this)
    }

}